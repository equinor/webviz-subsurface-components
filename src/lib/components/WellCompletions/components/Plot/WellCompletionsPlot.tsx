import { createStyles, makeStyles, Theme } from "@material-ui/core";
import React, { useMemo } from "react";
import { useResizeDetector } from "react-resize-detector";
import { PlotData } from "../../utils/dataUtil";
import { TooltipProvider } from "../Common/TooltipProvider";
import CompletionsPlot from "./CompletionsPlot";
import { getLayout, Padding } from "./plotUtil";
import StratigraphyPlot from "./StratigraphyPlot";
import WellsPlot from "./WellsPlot";

const useStyles = makeStyles((theme: Theme) =>
    createStyles({
        root: {
            display: "flex",
            flex: 1,
            height: "80%",
            padding: theme.spacing(1),
        },
    })
);
interface Props {
    timeSteps: string[];
    plotData: PlotData;
}

const padding: Padding = { left: 80, right: 50, top: 70, bottom: 50 };
/* eslint-disable react/prop-types */
const WellCompletionsPlot: React.FC<Props> = React.memo(
    ({ timeSteps, plotData }) => {
        const classes = useStyles();
        const { width, height, ref } = useResizeDetector({
            refreshMode: "debounce",
            refreshRate: 50,
            refreshOptions: { trailing: true },
        });
        const layout = useMemo(
            () =>
                width !== undefined && height !== undefined
                    ? getLayout(width, height, padding)
                    : undefined,
            [width, height]
        );

        return (
            <TooltipProvider>
                <div
                    className={classes.root}
                    ref={ref as React.LegacyRef<HTMLDivElement>}
                    data-tip
                    data-for="plot-tooltip"
                >
                    {layout && (
                        <svg
                            id={"svg-context"}
                            width={width}
                            height={height}
                            style={{ position: "relative" }}
                        >
                            <StratigraphyPlot
                                data={plotData.stratigraphy}
                                layout={layout}
                                padding={padding}
                            />
                            <WellsPlot
                                timeSteps={timeSteps}
                                plotData={plotData}
                                layout={layout}
                                padding={padding}
                            />
                            <CompletionsPlot
                                plotData={plotData}
                                layout={layout}
                                padding={padding}
                            />
                        </svg>
                    )}
                </div>
            </TooltipProvider>
        );
    }
);

export default WellCompletionsPlot;
