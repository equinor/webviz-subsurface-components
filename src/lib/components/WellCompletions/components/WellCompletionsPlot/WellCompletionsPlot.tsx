import { createStyles, makeStyles } from "@material-ui/core";
import { isEqual, throttle } from "lodash";
import React, { useEffect, useMemo, useRef } from "react";
import { useSelector } from "react-redux";
import ReactResizeDetector from "react-resize-detector";
import { WellCompletionsState } from "../../redux/store";
import { D3WellCompletions } from "./D3WellCompletions";
import { dataInTimeIndexRange } from "./dataUtil";

const useStyles = makeStyles(() =>
    createStyles({
        root: {
            display: "flex",
            flex: 1,
        },
        main: {
            width: "100%",
            height: "100%",
        },
    })
);
const WellCompletionsPlot: React.FC = () => {
    const classes = useStyles();

    // A reference to the div storing the plots
    const divRef = useRef<HTMLDivElement>();
    const d3wellcompletions = useRef<D3WellCompletions>();
    //States
    const id = useSelector((state: WellCompletionsState) => state.id);
    const data = useSelector(
        (state: WellCompletionsState) => state.dataModel.data!
    );
    const timeIndexRange = useSelector(
        (state: WellCompletionsState) => state.ui.timeIndexRange,
        isEqual
    ) as [number, number];
    //Memo
    const wellPlotData = useMemo(
        () => dataInTimeIndexRange(data, timeIndexRange),
        [data, timeIndexRange]
    );
    const onResize = useMemo(
        () =>
            throttle(
                () => {
                    if (d3wellcompletions.current) {
                        d3wellcompletions.current.resize();
                    }
                },
                100,
                { trailing: true }
            ),
        []
    );

    // Effects
    useEffect(() => {
        if (!d3wellcompletions.current) {
            d3wellcompletions.current = new D3WellCompletions(
                id,
                divRef.current as HTMLDivElement
            );
        }
    }, [id]);
    //Data changed
    useEffect(() => {
        if (d3wellcompletions.current) {
            d3wellcompletions.current.setStratigraphyData(data.stratigraphy);
        }
    }, [data.stratigraphy]);
    useEffect(() => {
        if (d3wellcompletions.current) {
            d3wellcompletions.current.setWellPlotData(wellPlotData);
        }
    }, [wellPlotData]);
    // on unmount
    useEffect(() => {
        return () => d3wellcompletions.current?.clear();
    }, []);

    return (
        <ReactResizeDetector handleHeight handleWidth onResize={onResize}>
            <div className={classes.root}>
                <div
                    className={classes.main}
                    ref={divRef as React.MutableRefObject<HTMLDivElement>}
                />
            </div>
        </ReactResizeDetector>
    );
};

export default WellCompletionsPlot;
