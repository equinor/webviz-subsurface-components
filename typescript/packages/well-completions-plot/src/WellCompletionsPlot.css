.WellCompletionsPlot {
    display: flex;
    flex: 1;
    height: 80%;
    padding: 8px;
}
