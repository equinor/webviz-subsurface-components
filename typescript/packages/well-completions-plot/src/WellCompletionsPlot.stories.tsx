import type { Meta, StoryObj } from "@storybook/react";
import React from "react";

import { WellCompletionsPlot } from "./WellCompletionsPlot";
import type { PlotData } from "./index";

import {
    timeSteps,
    firstPlotData,
    secondPlotData,
    thirdPlotData,
} from "../example-data/well-completions-plot-data";

/**
 * Storybook test for the well completions plot component
 */
const stories: Meta = {
    component: WellCompletionsPlot,
    title: "WellCompletionsPlot/Demo",
    argTypes: {
        data: {
            control: {
                type: "object",
            },
        },
    },
};
export default stories;

const Template = (data) => {
    const [plotData, setPlotData] = React.useState<PlotData>(data.plotData);
    const [timeSteps, setTimeSteps] = React.useState<string[]>(data.timeSteps);

    const [prevData, setPrevData] = React.useState(data);
    if (data !== prevData) {
        setPrevData(data);
        setPlotData(data.plotData);
        setTimeSteps(data.timeSteps);
    }

    const handleFirstButtonClick = () => {
        setPlotData(firstPlotData);
    };
    const handleSecondButtonClick = () => {
        setPlotData(secondPlotData);
    };
    const handleThirdButtonClick = () => {
        setPlotData(thirdPlotData);
    };

    return (
        <>
            <button type="button" onClick={() => handleFirstButtonClick()}>
                First plot data
            </button>
            <button type="button" onClick={() => handleSecondButtonClick()}>
                Second plot data
            </button>
            <button type="button" onClick={() => handleThirdButtonClick()}>
                Third plot data
            </button>
            <div
                style={{
                    minWidth: "500px",
                    minHeight: "300px",
                    height: "100vh",
                }}
            >
                <WellCompletionsPlot
                    id={"test"}
                    plotData={plotData}
                    timeSteps={timeSteps}
                />
            </div>
        </>
    );
};
export const WellCompletionsPlotStory: StoryObj<typeof Template> = {
    args: {
        plotData: firstPlotData,
        timeSteps: timeSteps,
    },
    render: (args) => <Template {...args} />,
};
