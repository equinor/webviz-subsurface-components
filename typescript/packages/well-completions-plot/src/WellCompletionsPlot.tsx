import React from "react";
import { useResizeDetector } from "react-resize-detector";
import { TooltipProvider } from "./components/TooltipProvider";
import { CompletionsPlot } from "./components/CompletionsPlot";
import type { Padding } from "./types/layoutTypes";
import { createLayout } from "./utils/layoutUtils";
import { StratigraphyPlot } from "./components/StratigraphyPlot";
import { WellsPlot } from "./components/WellsPlot";

import type { PlotData } from "./types/dataTypes";

import "./WellCompletionsPlot.css";

export interface WellCompletionsPlotProps {
    id: string;
    timeSteps: string[];
    plotData: PlotData;
}

const padding: Padding = { left: 80, right: 50, top: 70, bottom: 50 };
/* eslint-disable react/display-name */
/* eslint-disable react/prop-types */
export const WellCompletionsPlot: React.FC<WellCompletionsPlotProps> =
    React.memo((props: WellCompletionsPlotProps) => {
        const { width, height, ref } = useResizeDetector({
            refreshMode: "debounce",
            refreshRate: 50,
            refreshOptions: { trailing: true },
        });

        const layout =
            width !== undefined && height !== undefined
                ? createLayout(width, height, padding)
                : undefined;

        return (
            <TooltipProvider>
                <div
                    ref={ref as React.LegacyRef<HTMLDivElement>}
                    id={props.id}
                    className={"WellCompletionsPlot"}
                    data-tooltip-id="plot-tooltip"
                    data-tooltip-float={true}
                >
                    {layout && props.plotData && (
                        <svg
                            id={`${props.id}-svg-context`}
                            width={width}
                            height={height}
                            style={{ position: "relative" }}
                        >
                            <StratigraphyPlot
                                data={props.plotData.stratigraphy}
                                layout={layout}
                                padding={padding}
                            />
                            <WellsPlot
                                timeSteps={props.timeSteps}
                                plotData={props.plotData}
                                layout={layout}
                                padding={padding}
                            />
                            {props.plotData && (
                                <CompletionsPlot
                                    plotData={props.plotData}
                                    layout={layout}
                                    padding={padding}
                                />
                            )}
                        </svg>
                    )}
                </div>
            </TooltipProvider>
        );
    });

WellCompletionsPlot.displayName = "WellCompletionsPlot";
