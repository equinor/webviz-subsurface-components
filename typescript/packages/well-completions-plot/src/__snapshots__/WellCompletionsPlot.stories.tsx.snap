// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WellCompletionsPlot/Demo WellCompletionsPlotStory smoke-test: wellcompletionsplot-demo--well-completions-plot-story.dom.html 1`] = `
<button type="button">
  First plot data
</button>
<button type="button">
  Second plot data
</button>
<button type="button">
  Third plot data
</button>
<div style="min-width: 500px; min-height: 300px; height: 100vh;">
  <div id="test"
       class="WellCompletionsPlot"
       data-tooltip-id="plot-tooltip"
       data-tooltip-float="true"
  >
    <svg id="test-svg-context"
         width="1232"
         height="576"
         style="position: relative;"
    >
      <g>
        <g transform="translate(0,70)">
          <rect transform="translate(80, 0)"
                width="1102"
                height="32.57142857142857"
                fill="#808080"
          >
          </rect>
          <title>
            zone1.1
          </title>
          <text text-anchor="end"
                x="76"
                y="16.285714285714285"
                dy=".35em"
                font-family="sans-serif"
                style="font-size: 11px;"
          >
            zone1.1
          </text>
        </g>
        <g transform="translate(0,102.57142857142857)">
          <rect transform="translate(80, 0)"
                width="1102"
                height="32.57142857142857"
                fill="#FF0000"
          >
          </rect>
          <title>
            zone1.2
          </title>
          <text text-anchor="end"
                x="76"
                y="16.285714285714285"
                dy=".35em"
                font-family="sans-serif"
                style="font-size: 11px;"
          >
            zone1.2
          </text>
        </g>
        <g transform="translate(0,135.14285714285714)">
          <rect transform="translate(80, 0)"
                width="1102"
                height="32.57142857142857"
                fill="#800000"
          >
          </rect>
          <title>
            zone1.3
          </title>
          <text text-anchor="end"
                x="76"
                y="16.285714285714285"
                dy=".35em"
                font-family="sans-serif"
                style="font-size: 11px;"
          >
            zone1.3
          </text>
        </g>
        <g transform="translate(0,167.71428571428572)">
          <rect transform="translate(80, 0)"
                width="1102"
                height="32.57142857142857"
                fill="#FFFF00"
          >
          </rect>
          <title>
            zone2.1
          </title>
          <text text-anchor="end"
                x="76"
                y="16.285714285714285"
                dy=".35em"
                font-family="sans-serif"
                style="font-size: 11px;"
          >
            zone2.1
          </text>
        </g>
        <g transform="translate(0,200.28571428571428)">
          <rect transform="translate(80, 0)"
                width="1102"
                height="32.57142857142857"
                fill="#808000"
          >
          </rect>
          <title>
            zone2.2
          </title>
          <text text-anchor="end"
                x="76"
                y="16.285714285714285"
                dy=".35em"
                font-family="sans-serif"
                style="font-size: 11px;"
          >
            zone2.2
          </text>
        </g>
        <g transform="translate(0,232.85714285714283)">
          <rect transform="translate(80, 0)"
                width="1102"
                height="32.57142857142857"
                fill="#00FF00"
          >
          </rect>
          <title>
            zone2.3
          </title>
          <text text-anchor="end"
                x="76"
                y="16.285714285714285"
                dy=".35em"
                font-family="sans-serif"
                style="font-size: 11px;"
          >
            zone2.3
          </text>
        </g>
        <g transform="translate(0,265.42857142857144)">
          <rect transform="translate(80, 0)"
                width="1102"
                height="32.57142857142857"
                fill="#008000"
          >
          </rect>
          <title>
            zone3.1
          </title>
          <text text-anchor="end"
                x="76"
                y="16.285714285714285"
                dy=".35em"
                font-family="sans-serif"
                style="font-size: 11px;"
          >
            zone3.1
          </text>
        </g>
        <g transform="translate(0,298)">
          <rect transform="translate(80, 0)"
                width="1102"
                height="32.57142857142857"
                fill="#00FFFF"
          >
          </rect>
          <title>
            zone3.2
          </title>
          <text text-anchor="end"
                x="76"
                y="16.285714285714285"
                dy=".35em"
                font-family="sans-serif"
                style="font-size: 11px;"
          >
            zone3.2
          </text>
        </g>
        <g transform="translate(0,330.57142857142856)">
          <rect transform="translate(80, 0)"
                width="1102"
                height="32.57142857142857"
                fill="#008080"
          >
          </rect>
          <title>
            zone3.3
          </title>
          <text text-anchor="end"
                x="76"
                y="16.285714285714285"
                dy=".35em"
                font-family="sans-serif"
                style="font-size: 11px;"
          >
            zone3.3
          </text>
        </g>
        <g transform="translate(0,363.1428571428571)">
          <rect transform="translate(80, 0)"
                width="1102"
                height="32.57142857142857"
                fill="#0000FF"
          >
          </rect>
          <title>
            zone4.1
          </title>
          <text text-anchor="end"
                x="76"
                y="16.285714285714285"
                dy=".35em"
                font-family="sans-serif"
                style="font-size: 11px;"
          >
            zone4.1
          </text>
        </g>
        <g transform="translate(0,395.71428571428567)">
          <rect transform="translate(80, 0)"
                width="1102"
                height="32.57142857142857"
                fill="#000080"
          >
          </rect>
          <title>
            zone4.2
          </title>
          <text text-anchor="end"
                x="76"
                y="16.285714285714285"
                dy=".35em"
                font-family="sans-serif"
                style="font-size: 11px;"
          >
            zone4.2
          </text>
        </g>
        <g transform="translate(0,428.2857142857143)">
          <rect transform="translate(80, 0)"
                width="1102"
                height="32.57142857142857"
                fill="#FF00FF"
          >
          </rect>
          <title>
            zone4.3
          </title>
          <text text-anchor="end"
                x="76"
                y="16.285714285714285"
                dy=".35em"
                font-family="sans-serif"
                style="font-size: 11px;"
          >
            zone4.3
          </text>
        </g>
        <g transform="translate(0,460.85714285714283)">
          <rect transform="translate(80, 0)"
                width="1102"
                height="32.57142857142857"
                fill="#800080"
          >
          </rect>
          <title>
            zone4.4
          </title>
          <text text-anchor="end"
                x="76"
                y="16.285714285714285"
                dy=".35em"
                font-family="sans-serif"
                style="font-size: 11px;"
          >
            zone4.4
          </text>
        </g>
        <g transform="translate(0,493.4285714285714)">
          <rect transform="translate(80, 0)"
                width="1102"
                height="32.57142857142857"
                fill="#C0C0C0"
          >
          </rect>
          <title>
            zone4.5
          </title>
          <text text-anchor="end"
                x="76"
                y="16.285714285714285"
                dy=".35em"
                font-family="sans-serif"
                style="font-size: 11px;"
          >
            zone4.5
          </text>
        </g>
      </g>
      <g>
        <g transform="translate(171.83333333333331,0)">
          <text text-anchor="start"
                transform="translate(0,60) rotate(-60)"
                x="0"
                y="0"
                dy=".35em"
                font-family="sans-serif"
                style="font-size: 9px;"
          >
            PP_1
          </text>
          <rect transform="translate(0,66)"
                width="0.5"
                height="329.71428571428567"
                fill="#111"
          >
          </rect>
        </g>
        <g transform="translate(355.5,0)">
          <text text-anchor="start"
                transform="translate(0,60) rotate(-60)"
                x="0"
                y="0"
                dy=".35em"
                font-family="sans-serif"
                style="font-size: 9px;"
          >
            OP_2
          </text>
          <rect transform="translate(0,66)"
                width="0.5"
                height="362.2857142857143"
                fill="#111"
          >
          </rect>
        </g>
        <g transform="translate(539.1666666666666,0)">
          <text text-anchor="start"
                transform="translate(0,60) rotate(-60)"
                x="0"
                y="0"
                dy=".35em"
                font-family="sans-serif"
                style="font-size: 9px;"
          >
            OP_3
          </text>
          <rect transform="translate(0,66)"
                width="0.5"
                height="460"
                fill="#111"
          >
          </rect>
        </g>
        <g transform="translate(722.8333333333333,0)">
          <text text-anchor="start"
                transform="translate(0,60) rotate(-60)"
                x="0"
                y="0"
                dy=".35em"
                font-family="sans-serif"
                style="font-size: 9px;"
          >
            RWI_2
          </text>
          <rect transform="translate(0,66)"
                width="0.5"
                height="4"
                fill="#111"
          >
          </rect>
        </g>
        <g transform="translate(906.5,0)">
          <text text-anchor="start"
                transform="translate(0,60) rotate(-60)"
                x="0"
                y="0"
                dy=".35em"
                font-family="sans-serif"
                style="font-size: 9px;"
          >
            WI_1
          </text>
          <rect transform="translate(0,66)"
                width="0.5"
                height="166.85714285714283"
                fill="#111"
          >
          </rect>
        </g>
        <g transform="translate(1090.1666666666665,0)">
          <text text-anchor="start"
                transform="translate(0,60) rotate(-60)"
                x="0"
                y="0"
                dy=".35em"
                font-family="sans-serif"
                style="font-size: 9px;"
          >
            WI_2
          </text>
          <rect transform="translate(0,66)"
                width="0.5"
                height="4"
                fill="#111"
          >
          </rect>
        </g>
      </g>
      <g>
        <g transform="translate(171.83333333333331, 0)">
          <g>
            <rect transform="translate(0, 70)"
                  width="0"
                  height="130.28571428571428"
                  fill="#f00"
            >
            </rect>
            <rect transform="translate(0, 70)"
                  width="0"
                  height="130.28571428571428"
                  fill="#111"
            >
            </rect>
            <rect transform="translate(0, 70)"
                  width="0"
                  height="130.28571428571428"
                  fill="#f00"
            >
            </rect>
          </g>
          <g>
            <rect transform="translate(-45.916666666666664, 200.28571428571428)"
                  width="0"
                  height="32.57142857142857"
                  fill="#f00"
            >
            </rect>
            <rect transform="translate(-45.916666666666664, 200.28571428571428)"
                  width="91.83333333333333"
                  height="32.57142857142857"
                  fill="#111"
            >
            </rect>
            <rect transform="translate(45.916666666666664, 200.28571428571428)"
                  width="0"
                  height="32.57142857142857"
                  fill="#f00"
            >
            </rect>
          </g>
          <g>
            <rect transform="translate(-45.916666666666664, 232.85714285714283)"
                  width="0"
                  height="32.57142857142857"
                  fill="#f00"
            >
            </rect>
            <rect transform="translate(-45.916666666666664, 232.85714285714283)"
                  width="91.83333333333333"
                  height="32.57142857142857"
                  fill="#111"
            >
            </rect>
            <rect transform="translate(45.916666666666664, 232.85714285714283)"
                  width="0"
                  height="32.57142857142857"
                  fill="#f00"
            >
            </rect>
          </g>
          <g>
            <rect transform="translate(-45.916666666666664, 265.42857142857144)"
                  width="0"
                  height="32.57142857142857"
                  fill="#f00"
            >
            </rect>
            <rect transform="translate(-45.916666666666664, 265.42857142857144)"
                  width="91.83333333333333"
                  height="32.57142857142857"
                  fill="#111"
            >
            </rect>
            <rect transform="translate(45.916666666666664, 265.42857142857144)"
                  width="0"
                  height="32.57142857142857"
                  fill="#f00"
            >
            </rect>
          </g>
          <g>
            <rect transform="translate(-45.916666666666664, 298)"
                  width="0"
                  height="32.57142857142857"
                  fill="#f00"
            >
            </rect>
            <rect transform="translate(-45.916666666666664, 298)"
                  width="91.83333333333333"
                  height="32.57142857142857"
                  fill="#111"
            >
            </rect>
            <rect transform="translate(45.916666666666664, 298)"
                  width="0"
                  height="32.57142857142857"
                  fill="#f00"
            >
            </rect>
          </g>
          <g>
            <rect transform="translate(-45.916666666666664, 330.57142857142856)"
                  width="0"
                  height="32.57142857142857"
                  fill="#f00"
            >
            </rect>
            <rect transform="translate(-45.916666666666664, 330.57142857142856)"
                  width="91.83333333333333"
                  height="32.57142857142857"
                  fill="#111"
            >
            </rect>
            <rect transform="translate(45.916666666666664, 330.57142857142856)"
                  width="0"
                  height="32.57142857142857"
                  fill="#f00"
            >
            </rect>
          </g>
          <g>
            <rect transform="translate(-45.916666666666664, 363.1428571428571)"
                  width="0"
                  height="32.57142857142857"
                  fill="#f00"
            >
            </rect>
            <rect transform="translate(-45.916666666666664, 363.1428571428571)"
                  width="91.83333333333333"
                  height="32.57142857142857"
                  fill="#111"
            >
            </rect>
            <rect transform="translate(45.916666666666664, 363.1428571428571)"
                  width="0"
                  height="32.57142857142857"
                  fill="#f00"
            >
            </rect>
          </g>
          <g>
            <rect transform="translate(0, 395.71428571428567)"
                  width="0"
                  height="130.28571428571428"
                  fill="#f00"
            >
            </rect>
            <rect transform="translate(0, 395.71428571428567)"
                  width="0"
                  height="130.28571428571428"
                  fill="#111"
            >
            </rect>
            <rect transform="translate(0, 395.71428571428567)"
                  width="0"
                  height="130.28571428571428"
                  fill="#f00"
            >
            </rect>
          </g>
        </g>
        <g transform="translate(355.5, 0)">
          <g>
            <rect transform="translate(0, 70)"
                  width="0"
                  height="195.42857142857142"
                  fill="#f00"
            >
            </rect>
            <rect transform="translate(0, 70)"
                  width="0"
                  height="195.42857142857142"
                  fill="#111"
            >
            </rect>
            <rect transform="translate(0, 70)"
                  width="0"
                  height="195.42857142857142"
                  fill="#f00"
            >
            </rect>
          </g>
          <g>
            <rect transform="translate(-45.916666666666664, 265.42857142857144)"
                  width="0"
                  height="32.57142857142857"
                  fill="#f00"
            >
            </rect>
            <rect transform="translate(-45.916666666666664, 265.42857142857144)"
                  width="91.83333333333333"
                  height="32.57142857142857"
                  fill="#111"
            >
            </rect>
            <rect transform="translate(45.916666666666664, 265.42857142857144)"
                  width="0"
                  height="32.57142857142857"
                  fill="#f00"
            >
            </rect>
          </g>
          <g>
            <rect transform="translate(-45.916666666666664, 298)"
                  width="0"
                  height="32.57142857142857"
                  fill="#f00"
            >
            </rect>
            <rect transform="translate(-45.916666666666664, 298)"
                  width="91.83333333333333"
                  height="32.57142857142857"
                  fill="#111"
            >
            </rect>
            <rect transform="translate(45.916666666666664, 298)"
                  width="0"
                  height="32.57142857142857"
                  fill="#f00"
            >
            </rect>
          </g>
          <g>
            <rect transform="translate(-45.916666666666664, 330.57142857142856)"
                  width="0"
                  height="32.57142857142857"
                  fill="#f00"
            >
            </rect>
            <rect transform="translate(-45.916666666666664, 330.57142857142856)"
                  width="91.83333333333333"
                  height="32.57142857142857"
                  fill="#111"
            >
            </rect>
            <rect transform="translate(45.916666666666664, 330.57142857142856)"
                  width="0"
                  height="32.57142857142857"
                  fill="#f00"
            >
            </rect>
          </g>
          <g>
            <rect transform="translate(-45.916666666666664, 363.1428571428571)"
                  width="0"
                  height="32.57142857142857"
                  fill="#f00"
            >
            </rect>
            <rect transform="translate(-45.916666666666664, 363.1428571428571)"
                  width="91.83333333333333"
                  height="32.57142857142857"
                  fill="#111"
            >
            </rect>
            <rect transform="translate(45.916666666666664, 363.1428571428571)"
                  width="0"
                  height="32.57142857142857"
                  fill="#f00"
            >
            </rect>
          </g>
          <g>
            <rect transform="translate(-45.916666666666664, 395.71428571428567)"
                  width="0"
                  height="32.57142857142857"
                  fill="#f00"
            >
            </rect>
            <rect transform="translate(-45.916666666666664, 395.71428571428567)"
                  width="91.83333333333333"
                  height="32.57142857142857"
                  fill="#111"
            >
            </rect>
            <rect transform="translate(45.916666666666664, 395.71428571428567)"
                  width="0"
                  height="32.57142857142857"
                  fill="#f00"
            >
            </rect>
          </g>
          <g>
            <rect transform="translate(0, 428.2857142857143)"
                  width="0"
                  height="97.71428571428571"
                  fill="#f00"
            >
            </rect>
            <rect transform="translate(0, 428.2857142857143)"
                  width="0"
                  height="97.71428571428571"
                  fill="#111"
            >
            </rect>
            <rect transform="translate(0, 428.2857142857143)"
                  width="0"
                  height="97.71428571428571"
                  fill="#f00"
            >
            </rect>
          </g>
        </g>
        <g transform="translate(539.1666666666666, 0)">
          <g>
            <rect transform="translate(0, 70)"
                  width="0"
                  height="260.57142857142856"
                  fill="#f00"
            >
            </rect>
            <rect transform="translate(0, 70)"
                  width="0"
                  height="260.57142857142856"
                  fill="#111"
            >
            </rect>
            <rect transform="translate(0, 70)"
                  width="0"
                  height="260.57142857142856"
                  fill="#f00"
            >
            </rect>
          </g>
          <g>
            <rect transform="translate(-45.916666666666664, 330.57142857142856)"
                  width="0"
                  height="32.57142857142857"
                  fill="#f00"
            >
            </rect>
            <rect transform="translate(-45.916666666666664, 330.57142857142856)"
                  width="91.83333333333333"
                  height="32.57142857142857"
                  fill="#111"
            >
            </rect>
            <rect transform="translate(45.916666666666664, 330.57142857142856)"
                  width="0"
                  height="32.57142857142857"
                  fill="#f00"
            >
            </rect>
          </g>
          <g>
            <rect transform="translate(-45.916666666666664, 363.1428571428571)"
                  width="0"
                  height="32.57142857142857"
                  fill="#f00"
            >
            </rect>
            <rect transform="translate(-45.916666666666664, 363.1428571428571)"
                  width="91.83333333333333"
                  height="32.57142857142857"
                  fill="#111"
            >
            </rect>
            <rect transform="translate(45.916666666666664, 363.1428571428571)"
                  width="0"
                  height="32.57142857142857"
                  fill="#f00"
            >
            </rect>
          </g>
          <g>
            <rect transform="translate(-45.916666666666664, 395.71428571428567)"
                  width="0"
                  height="32.57142857142857"
                  fill="#f00"
            >
            </rect>
            <rect transform="translate(-45.916666666666664, 395.71428571428567)"
                  width="91.83333333333333"
                  height="32.57142857142857"
                  fill="#111"
            >
            </rect>
            <rect transform="translate(45.916666666666664, 395.71428571428567)"
                  width="0"
                  height="32.57142857142857"
                  fill="#f00"
            >
            </rect>
          </g>
          <g>
            <rect transform="translate(-45.916666666666664, 428.2857142857143)"
                  width="0"
                  height="32.57142857142857"
                  fill="#f00"
            >
            </rect>
            <rect transform="translate(-45.916666666666664, 428.2857142857143)"
                  width="91.83333333333333"
                  height="32.57142857142857"
                  fill="#111"
            >
            </rect>
            <rect transform="translate(45.916666666666664, 428.2857142857143)"
                  width="0"
                  height="32.57142857142857"
                  fill="#f00"
            >
            </rect>
          </g>
          <g>
            <rect transform="translate(-45.916666666666664, 460.85714285714283)"
                  width="0"
                  height="32.57142857142857"
                  fill="#f00"
            >
            </rect>
            <rect transform="translate(-45.916666666666664, 460.85714285714283)"
                  width="91.83333333333333"
                  height="32.57142857142857"
                  fill="#111"
            >
            </rect>
            <rect transform="translate(45.916666666666664, 460.85714285714283)"
                  width="0"
                  height="32.57142857142857"
                  fill="#f00"
            >
            </rect>
          </g>
          <g>
            <rect transform="translate(-45.916666666666664, 493.4285714285714)"
                  width="0"
                  height="32.57142857142857"
                  fill="#f00"
            >
            </rect>
            <rect transform="translate(-45.916666666666664, 493.4285714285714)"
                  width="91.83333333333333"
                  height="32.57142857142857"
                  fill="#111"
            >
            </rect>
            <rect transform="translate(45.916666666666664, 493.4285714285714)"
                  width="0"
                  height="32.57142857142857"
                  fill="#f00"
            >
            </rect>
          </g>
        </g>
        <g transform="translate(722.8333333333333, 0)">
          <g>
            <rect transform="translate(0, 70)"
                  width="0"
                  height="456"
                  fill="#f00"
            >
            </rect>
            <rect transform="translate(0, 70)"
                  width="0"
                  height="456"
                  fill="#111"
            >
            </rect>
            <rect transform="translate(0, 70)"
                  width="0"
                  height="456"
                  fill="#f00"
            >
            </rect>
          </g>
        </g>
        <g transform="translate(906.5, 0)">
          <g>
            <rect transform="translate(0, 70)"
                  width="0"
                  height="32.57142857142857"
                  fill="#f00"
            >
            </rect>
            <rect transform="translate(0, 70)"
                  width="0"
                  height="32.57142857142857"
                  fill="#111"
            >
            </rect>
            <rect transform="translate(0, 70)"
                  width="0"
                  height="32.57142857142857"
                  fill="#f00"
            >
            </rect>
          </g>
          <g>
            <rect transform="translate(-45.916666666666664, 102.57142857142857)"
                  width="0"
                  height="32.57142857142857"
                  fill="#f00"
            >
            </rect>
            <rect transform="translate(-45.916666666666664, 102.57142857142857)"
                  width="91.83333333333333"
                  height="32.57142857142857"
                  fill="#111"
            >
            </rect>
            <rect transform="translate(45.916666666666664, 102.57142857142857)"
                  width="0"
                  height="32.57142857142857"
                  fill="#f00"
            >
            </rect>
          </g>
          <g>
            <rect transform="translate(-45.916666666666664, 135.14285714285714)"
                  width="0"
                  height="32.57142857142857"
                  fill="#f00"
            >
            </rect>
            <rect transform="translate(-45.916666666666664, 135.14285714285714)"
                  width="91.83333333333333"
                  height="32.57142857142857"
                  fill="#111"
            >
            </rect>
            <rect transform="translate(45.916666666666664, 135.14285714285714)"
                  width="0"
                  height="32.57142857142857"
                  fill="#f00"
            >
            </rect>
          </g>
          <g>
            <rect transform="translate(-45.916666666666664, 167.71428571428572)"
                  width="0"
                  height="32.57142857142857"
                  fill="#f00"
            >
            </rect>
            <rect transform="translate(-45.916666666666664, 167.71428571428572)"
                  width="91.83333333333333"
                  height="32.57142857142857"
                  fill="#111"
            >
            </rect>
            <rect transform="translate(45.916666666666664, 167.71428571428572)"
                  width="0"
                  height="32.57142857142857"
                  fill="#f00"
            >
            </rect>
          </g>
          <g>
            <rect transform="translate(-45.916666666666664, 200.28571428571428)"
                  width="0"
                  height="32.57142857142857"
                  fill="#f00"
            >
            </rect>
            <rect transform="translate(-45.916666666666664, 200.28571428571428)"
                  width="91.83333333333333"
                  height="32.57142857142857"
                  fill="#111"
            >
            </rect>
            <rect transform="translate(45.916666666666664, 200.28571428571428)"
                  width="0"
                  height="32.57142857142857"
                  fill="#f00"
            >
            </rect>
          </g>
          <g>
            <rect transform="translate(0, 232.85714285714283)"
                  width="0"
                  height="293.1428571428571"
                  fill="#f00"
            >
            </rect>
            <rect transform="translate(0, 232.85714285714283)"
                  width="0"
                  height="293.1428571428571"
                  fill="#111"
            >
            </rect>
            <rect transform="translate(0, 232.85714285714283)"
                  width="0"
                  height="293.1428571428571"
                  fill="#f00"
            >
            </rect>
          </g>
        </g>
        <g transform="translate(1090.1666666666665, 0)">
          <g>
            <rect transform="translate(0, 70)"
                  width="0"
                  height="456"
                  fill="#f00"
            >
            </rect>
            <rect transform="translate(0, 70)"
                  width="0"
                  height="456"
                  fill="#111"
            >
            </rect>
            <rect transform="translate(0, 70)"
                  width="0"
                  height="456"
                  fill="#f00"
            >
            </rect>
          </g>
        </g>
      </g>
    </svg>
  </div>
</div>
`;
